feat(ugc): self-serve listings purchase block

dkoo opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Implements a purchase form block for self-serve listings. This PR only implements the block in the editor and its front-end output, not the handling of the form submission (which will be in a separate PR). The look and behavior of the block are heavily inspired by the Donate block in Newspack Blocks, both for ease of feature development and to maintain a consistent UX across Newspack products.

The form is designed to contain as few fields as possible in order to reduce purchase friction. I'm hoping that the actual content editing of purchased listings can happen directly in the block editor, but more to come on that in a future PR. I may add/remove/change the inputs as needed as we develop the full purchase flow.

@thomasguillot and/or @laurelfulford, feel free to suggest design-related changes or just push your own changes as needed.

As with all purchase-related self-serve features, this block is gated behind the `NEWSPACK_LISTINGS_SELF_SERVE_ENABLED` environment flag, so if this flag isn't set to `true` in `wp-config.php`, it won't be available.

<img width="785" alt="Screen Shot 2021-09-24 at 3 30 09 PM" src="https://user-images.githubusercontent.com/2230142/134741843-e4ca74ac-3752-4cbe-a665-12a5afbc4756.png">

### How to test the changes in this Pull Request:

1. In `wp-config.php`, set `define( 'NEWSPACK_LISTINGS_SELF_SERVE_ENABLED', true );` to enable the feature.
2. In a post or page, confirm there's a new block in the block inserter under the "Newspack" category:

<img width="340" alt="Screen Shot 2021-09-24 at 3 33 34 PM" src="https://user-images.githubusercontent.com/2230142/134742026-de37f634-d12c-4ab4-b1b8-09487516c412.png">

3. Insert the block into your content. Note that in the editor, the form is interactive but you can't submit it. You can edit certain copy within the block preview, such as the descriptions for single and subscription products, and the Submit button text. Play with the the block attributes and see how the different attribute settings behave on the front-end. Make sure to preview at various viewports to ensure that the styles look good in all cases.
4. Confirm that when submitting the form on the front-end, it doesn't do anything at the moment, but it does append URL params according to the form values you enter (these will be intercepted and used to add the corresponding products to the WC cart before redirecting to Checkout, just like with the Donate flow). The URL should look something like:

```
<PAGE URL>/?listing-purchase-type=single&listing-single-type=classified&listing-featured-upgrade=on&listing-title=My+Self-Serve+Listing
```
5. In `wp-config.php`, set `define( 'NEWSPACK_LISTINGS_SELF_SERVE_ENABLED', false );` or delete the constant entirely to disable the feature.
6. Confirm that in the editor, the block is no longer insertable (and the block you previously inserted shows a "not supported" error). Confirm on the front-end that the block renders no markup.
7. For kicks, you can re-enable the feature and activate a non-Newspack theme to confirm that the styles look surprisingly intact even without the Newspack CSS:

<img width="629" alt="Screen Shot 2021-09-24 at 3 42 21 PM" src="https://user-images.githubusercontent.com/2230142/134742741-dec76c87-f03e-4711-bb88-d306bae4e7f4.png">

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

thomasguillot commented:
I find the order of the various inputs a bit confusing. I'd suggest to start with the title, then the type, then the upgrade checkbox/toggle.

Can we also simplify the labels? It feel a bit long-winded.

![Screenshot 2021-09-27 at 11 41 45](https://user-images.githubusercontent.com/177929/134894003-ac898c21-abbd-4e9d-82be-b1d326abebcc.png)

dkoo commented:
> I find the order of the various inputs a bit confusing. I'd suggest to start with the title, then the type, then the upgrade checkbox/toggle.

@thomasguillot Totally understandable—I was trying to get away with using only one "listing title" field for both tabs 😄 

I updated the order and renamed the labels in 1c64b5f. The caveat is that I had to duplicate the Listing Title field in both tabs, so now they can no longer be `required` for form submission. Probably not a big deal as if there's no value for the title, we can create a listing with a placeholder title like "Untitled Listing".

claudiulodro reviewed (approved):
Looks solid! Great job! I didn't notice any issues in the code, and the design is easy to understand, especially for people with Donate block experience.
